src/main/kotlin/com/manimdsl/frontend/SymbolTable.kt

```diff
@@ -2,30 +2,51 @@ package com.manimdsl.frontend
 
 data class IdentifierData(val type: Type)
 
-interface SymbolTable {
-    fun getTypeOf(identifier: String): Type
-}
+class SymbolTable {
+    private val scopes = mutableListOf<ScopedSymbolTable>(GlobalScopeSymbolTable())
+    private var currentScope: ScopedSymbolTable = scopes[0]
+
+    fun getTypeOf(identifier: String): Type = currentScope.getTypeOf(identifier)
+
+    fun addVariable(identifier: String, type: Type) {
+        currentScope.addVariable(identifier, type)
+    }
+
+    fun enterScope(): Int {
+        val newScope = ScopedSymbolTable(currentScope)
+        scopes.add(newScope)
+        currentScope = newScope
+        // Returning index of this new symbol table (used in AST)
+        return scopes.size - 1
+    }
+
+    fun leaveScope() {
+        currentScope = currentScope.parent?: scopes[0]
+    }
 
-object SymbolTableRoot : SymbolTable {
-    override fun getTypeOf(identifier: String): Type {
-        return NoType
+    fun goToScope(scopeID: Int) {
+        if (scopeID in scopes.indices) {
+            currentScope = scopes[scopeID]
+        }
     }
+
+    fun getCurrentScopeID(): Int = scopes.indexOf(currentScope)
 }
 
-class SymbolTableNode : SymbolTable {
+class GlobalScopeSymbolTable: ScopedSymbolTable(null)
```

Comment:
Can this not be an object?

Reply:
Could mess up for tests due to concurrency issues an running all the tests at the same time

Reply:
Tests are not run in parallel

Reply:
No as in they will all use the same object and overwrite each other. I have tried to implement it as an object but it fails as leftover information from the other tests is persisted